File: archive/2026-04-05/ch06_tree/t10_count_nodes.py
"""LeetCode 222. Count Complete Tree Nodes
https://leetcode.cn/problems/count-complete-tree-nodes/
"""

from typing import Optional

from tree_node import TreeNode


class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        left, right, depth = root, root, 0
        while not left and not right:
            left = left.left
            right = right.right
            depth += 1

        if not left and not right:
            return 2 ** depth - 1
        else:
            return self.countNodes(root.left) + self.countNodes(root.right) + 1

if __name__ == "__main__":
    solution = Solution()
    root = TreeNode(
        1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), None)
    )
    assert solution.countNodes(root) == 6
    assert solution.countNodes(None) == 0
    print("All tests passed.")

File: archive/2025-04-04/ch06_tree/T10_countNodes.java
package org.review2025spring.ch06_tree;

public class T10_countNodes {
    public int countNodes(TreeNode root) {
        if (root == null) return 0;

        TreeNode left = root, right = root;
        int depth = 0;
        while (left != null && right != null) {
            left = left.left;
            right = right.right;
            depth++;
        }
        if (left == null && right==null)
            return (int) Math.pow(2, depth) - 1;

        return countNodes(root.left) + countNodes(root.right) + 1;
    }
}
